### PR TITLE
Add copy to clipboard button widget

### PR DESCRIPTION
**Pull request by raviks789**
The button copies the text from the specified target id when mentioned otherwise it copies the text from the parent element.

ref Icinga/icingadb-web#677

### CODE REVIEW

**Review by nilmerg — CHANGES_REQUESTED** [review]
This does not "resolve" the issue in Icinga DB Web. It refers to it at a maximum. Currently this looks like this: 
image

* Please only show the button upon hover of the parent (The rule for this must be available here!)
* Please fade "Copied" out
* The button's outline is placed weird, has different spacing left and right

File: asset/js/widget/CopyToClipboard.js [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+                } else {
+                    copyText = el.innerText;
+                }
+            } else {
+                copyText = button.parentElement.innerText;
+            }
+
+            navigator.clipboard.writeText(copyText).then(() => {
+                let previousHtml = button.innerHTML
+                button.innerHTML = 'Copied';
```
Comment (nilmerg) [inline asset/js/widget/CopyToClipboard.js:48]: Please fetch the label from a data attribute and translate it on the server.

File: asset/js/widget/CopyToClipboard.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
+define(["../notjQuery"], function ($) {
+
+    "use strict";
+
+    class CopyToClipboard {
+        constructor(button)
+        {
+            this.render(button);
+            $(button).on('click', null, this.onClick, this);
+
```
Comment (nilmerg) [inline asset/js/widget/CopyToClipboard.js:10]: please remove that empty line

**Review by nilmerg — CHANGES_REQUESTED** [review]

**Review by nilmerg — CHANGES_REQUESTED** [review]
I just noticed in the linked PR that we litter the markup with IDs for no real reason. The thing with IDs is that they are (need..) to be unique, but this uniqueness is not required here.

We already have a guaranteed wrapper. So why not use it as point of reference?

In the button's `parentElement` (which is assumed to be always the wrapper) we lookup a direct child with the attribute `data-clipboard-source` and use it as source.

No fallback needed. If it's not there, an error should be logged. (We have no source that has no id, at this time)

Anyone using `attachTo` doesn't need to adjust the element then in any way, the data-attribute can be set automatically.

File: src/Widget/CopyToClipboard.php [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        $clipboardWrapper = new HtmlElement(
+            'div',
+            Attributes::create(['class' => 'clipboard-wrapper'])
+        );
+
+        if ($source->hasAttribute('id')) {
+            $button->addAttributes(['data-clipboard-source' => $source->getAttribute('id')->getValue()]);
+        } else {
+            $button->addAttributes(['data-clipboard-source' => 'parent']);
+        }
```
Comment (nilmerg) [inline src/Widget/CopyToClipboard.php:62]: Remove this, the id of the source should be not of interest. Add the attribute to the source instead.

File: src/Widget/CopyToClipboard.php [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+        $clipboardWrapper->addHtml($source, $button);
+
+        return $clipboardWrapper;
+    }
+
+    public function assemble(): void
+    {
+        $this->setContent(new Icon('clone'));
```
Comment (nilmerg) [inline src/Widget/CopyToClipboard.php:71]: setHtmlContent?

File: src/Widget/CopyToClipboard.php [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+        if ($source->hasAttribute('id')) {
+            $button->addAttributes(['data-clipboard-source' => $source->getAttribute('id')->getValue()]);
+        } else {
+            $button->addAttributes(['data-clipboard-source' => 'parent']);
+        }
+
+        $clipboardWrapper->addHtml($source, $button);
+
+        return $clipboardWrapper;
```
Comment (nilmerg) [inline src/Widget/CopyToClipboard.php:66]: Please set the wrapper really as wrapper and return void here.

**Review by nilmerg — CHANGES_REQUESTED** [review]

File: asset/css/copy-to-clipboard.less [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+.copy-to-clipboard {
+  appearance: none;
+  border: none;
+  opacity: 0;
+  background: none;
+  padding: 0.25em 0.5em;
+
+  &:hover, &:focus {
+    opacity: 0.8;
+    color: @control-color;
```
Comment (nilmerg) [inline asset/css/copy-to-clipboard.less:26]:
```suggestion
    color: var(--control-color, @control-color);
```

**Review by nilmerg — CHANGES_REQUESTED** [review]
With the plan in mind to introduce the behavior for this in Icinga Web, the feature won't be available unless the user upgrades to the new release. The click on the button then has no effect. So please let the constructor of `CopyToClipboard.js` add a class that enables the hover/focus effect. And add a `tabindex=-1` by default to the button, which the constructor should also remove. This prevents the button from being reachable by keyboard.

**Review by flourish86 — APPROVED** [review]
Fine from the design perspective.

File: asset/js/widget/CopyToClipboard.js [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+                    button.innerText = button.dataset.copiedLabel;
+                    button.classList.add('persistent');
+
+                    // fade out after 3 second
+                    setTimeout(() => {
+                        // after 4 second, reset it.
+                        button.classList.add('copied');
+                        setTimeout(() => {
+                            button.classList.remove('copied');
+                            button.classList.remove('persistent');
```
Comment (nilmerg) [inline asset/js/widget/CopyToClipboard.js:37]: Having two classes for this is somehow redundant. Use only one please.

**Review by nilmerg — CHANGES_REQUESTED** [review]

**Review by nilmerg — APPROVED** [review]

**Pull request merged**